working/4271 - static enough metaprogramming/proposal.md

```diff
@@ -0,0 +1,1222 @@
+> runtime performance. This reduces the size of deployed applications but
+> decreases peak performance.
+
+There are multiple reasons for going this route. First of all, I believe that
+this allows to considerably simplify the design and implementation of this
+feature. But there is another reason for this: _you can't actually perform
+compile time evaluation without knowing compile time environment_. Consider for
+example the following code:
+
+```dart
+void foo(@konst String id) {
+  if (id.length != 10) {
+    throw 'Incorect value for $id';
+  }
+
+  // Do various things with id.
+}
+
+void bar() {
+  foo(const String.fromEnvironment('some.define'));
+}
+```
+
+Dart Analyzer can't really say what happens here because it does not know
+a specific value of `id` - the value is only known when compiling. The example
+can be made even more complicated by adding dependency on various platform
+specific defines (e.g. to distinguish between target platforms and specialize
+for them).
+
+Fundamentally this means that _compiling_ an application can reveal compile
+time error which are not revealed by running Dart analyzer on the same code.
+Furthermore compiling for different platforms can reveal different errors.
+
+The fact that these errors surface in compile time is an intrinsic property
+of this proposal - and is a direct consequence of its powerful ability to
+specialize code based on the compile time computation. It can't be fully
+avoided for the reasons which are explained above - though we could eventually
+_choose_ to implement subset of this proposal in analyzer as well and surface
+some of the errors earlier. However I currently don't see direct analyzer
+support as a necessary requirement to shipping this feature.
+
+> [!NOTE]
+>
+> If we decide that `@konst` is a no-op in debug (development) mode and
+> `@konst` reflection falls back to actual reflection then it is reasonable
+> to expect that developer could still opt-in into full `@konst` evaluation
+> for debug (development) mode. This is important to enable testing this code
+> without forcing developers to build their code in release mode.
+
+> [!NOTE]
+>
+> Some of our tools (most notably analyzer and DDC) are capable of
+> analyzing/compiling libraries only given the _outlines_ of their dependencies.
+> This was one of the showstoppers for enhanced constant proposal, which did not
+> introduce a syntactic marker to delineate code potentially needed for
+> `const`-evaluation from the rest of the program, which would mean that
+> outlines had to contain bodies for most methods - making them impractically
+> large and erasing their benefits.
+>
+> This proposal does not suffer the same issue, even if we decide to support
+> `@konst` evaluation in these tools: as `@konst` actually provides a syntactic
+> marker which partitions the program. Only bodies of methods with `@konst`
+> parameters need to be included into the outline.
```

Comment:
Can these functions only invoke other `@konst` functions?

Reply:
Yeah, I can see this happening. I also realized that what I am suggesting does not actually work: consider a simple class like this:

```dart
class Foo {
  final int v;
  const Foo(this.v);

  Foo inc(int u) => Foo(this.v + u);
}
```

There is nothing preventing us from evaluating something like `Foo(1).inc(2)` as constant - but we can't really mark `inc` as requiring constant inputs, because it does not require that. Not everything that can be constant evaluted should always be constant evaluted... Hence you can't use `@konst` as a marker.

Revised the text incorporating your suggestion that we probably should partition imports instead.